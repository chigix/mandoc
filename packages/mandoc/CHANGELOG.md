# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.2.1-alpha.0](https://github.com/chigix/mandoc/compare/mandoc@1.2.0...mandoc@1.2.1-alpha.0) (2018-12-08)


### Bug Fixes

* **mandoc:** fix import statement and path consts script in mandoc ([0a40605](https://github.com/chigix/mandoc/commit/0a40605))
