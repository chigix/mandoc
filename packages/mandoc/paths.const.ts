export const PKG_ROOT = __dirname;
export const TPL_DEFAULT_SRC_DIR = 'source';
export const TPL_DEFAULT_LAYOUT_FILE = 'layout/template.njk';
export const TPL_CFG_FILE_PATH = 'mandoc.config.js';
export const TPL_CFG_JSON_PATH = 'mandoc.config.json';
