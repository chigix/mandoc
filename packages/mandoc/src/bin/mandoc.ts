#!/usr/bin/env node

import chalk from 'chalk';
import * as program from 'commander';
import * as fs from 'fs';
// TODO: not suitable for compiling to esm5 esm2015 and types packages
import _ = require('lodash');
import * as path from 'path';
import * as stream from 'stream';
import * as through from 'through2';
import { TPL_PKG_ALIASES } from '../alias.const';
import { CmdMandocOptions, Path } from '../interfaces';
import { DocumentNotFoundError } from '../lib/errors';
import { extName, generatePrintContext } from '../lib/util';
import { PKG_ROOT } from '../paths.const';
import SiteWrapper from '../scripts/html2site.stream';
import { renderMarkdown } from '../scripts/md2html.stream';
import createPDFRenderStream from '../scripts/site2pdf.stream';
import { getConfig as templateConfigure } from '../scripts/template.config';

const pkg_json = require(path.resolve(PKG_ROOT, './package.json'));

program.version(pkg_json.version, '-v, --version')
  .description(pkg_json.description)
  .usage('[options] <FILE>')
  .option('<FILE>', 'Path of the markdown file to convert. '
    + 'If no source file specified, the stdin is to be used')
  .option('-o, --output [FILE]', 'Write output to FILE instead of stdout')
  .option('-w, --watch', 'Watch file for changes and rewrite')
  .option('-f, --from [FORMAT]', 'Specify input format. Markdown Default')
  .option('-t, --to [FORMAT]', 'Specify output format. PDF Default')
  .option('--template [NAME|DIR]',
    'Use a template for the generated document, `paper` package is default.')
  .option('--pagesize, --print.pageSize [SIZE]',
    'Paper Size, e.g. `letter`, `A4 landscape`')
  .option('--toc, --table-of-contents', 'non-null value if specified')
  .option('--toc-title', 'title of table of contents')
  .option('--lint', 'Lint Tool to check format and style issues in the document')
  // Planed in the future. Smart Quotes is needed for tex input improve
  // command.option('--smart', 'Interpret quotes, hyphens and ellipses');
  // Planed in the future. Only try to remove dependencies in markdown content.
  // External Dependencies in templates is not involved.
  // command.option('--self-contained',
  //   'Produce a standalone HTML file with no external dependencies');
  .option('--verbose', 'Give verbose debugging output')
  .action(async (
    file: Path | null,
    cmd: { [key: string]: string },
  ) => {
    if (!_.isString(file)) {
      cmd = file || {};
      file = null;
    }
    if (_.isString(cmd)) {
      process.stderr.write(chalk.redBright(`Unrecognized option: '${cmd}'`));
      process.exit(1);
    }

    const options = (function validateAndFixOptions() {
      const ext_name = _.isNull(file) ? '.md' : path.extname(file);
      const result = _.assign({
        from: ext_name.substring(1),
        to: 'pdf',
        template: TPL_PKG_ALIASES.default,
        tableOfContents: false,
        output: _.isString(file) ?
          file.substring(0, file.length - ext_name.length) :
          null,
        watch: false,
      } as CmdMandocOptions, cmd);
      if (result.to && result.output) {
        result.output = `${result.output}.${result.to}`;
      }
      if (!_.isString(result.output)) {
        process.stderr.write(chalk.redBright('Please provide output path'));
        process.exit(1);
      }

      return result;
    })();

    const start_stream = await new Promise<NodeJS.ReadableStream>(
      (resolve, reject) => {
        if (_.isNull(file)) {
          return resolve(process.stdin);
        }
        const file_path = file;
        fs.lstat(file_path, (err, stats) => {
          if (err) {
            return reject(err);
          }
          if (!stats.isFile()) {
            return reject(new DocumentNotFoundError(
              `Markdown Source File is not existed: ${file}`));
          }

          return resolve(fs.createReadStream(file_path));
        });
      }).catch((err) => {
        if (err.errno && err.code === 'ENOENT') {
          chalk.redBright('No such file or directory : ' + file);
        } else if (err instanceof DocumentNotFoundError) {
          chalk.redBright('No such file or directory : ' + file);
        } else {
          throw err;
        }
      });

    if (!start_stream) {
      return;
    }

    const PRINT_CTX = generatePrintContext(options['print.pageSize'] || 'A4');
    const TPL_CTX = templateConfigure(options.template, PRINT_CTX);

    let building_stream = start_stream.pipe(
      (
        ({
          'md': () => renderMarkdown({
            baseDir: file ? path.dirname(file) : process.cwd(),
          }),
          'markdown': () => renderMarkdown({
            baseDir: file ? path.dirname(file) : process.cwd(),
          }),
          // TODO: Tex Support here
        } as {
            [key: string]: (() => stream.Transform) | undefined,
          }
        )[options.from] || (() => through())
      )(),
    ).pipe(
      (function tplStream() {
        return TPL_CTX.renderers[extName(TPL_CTX.main)]
          || TPL_CTX.renderers.njk;
      })().stream.on('error', function (e) {
        chalk.redBright(e.stack as string);
      }),
    );
    if (options.to === 'pdf') {
      building_stream = building_stream.pipe(
        SiteWrapper(TPL_CTX),
      ).pipe(
        createPDFRenderStream(TPL_CTX, PRINT_CTX),
      );
    }
    building_stream.pipe(
      fs.createWriteStream(options.output),
    ).on('finish', () => {
      process.exit(0);
    }).on('error', (e) => {
      process.stderr.write(chalk.redBright(e.message));
      process.exit(1);
    });
  });

(function main() {
  if (!process.argv.slice(2).length) {
    program.outputHelp();
    process.exit();
  }

  program.parse(process.argv);
})();
