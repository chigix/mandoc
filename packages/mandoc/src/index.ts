export {
  LESS_STREAM_FACTORY as createLessCompileStream,
  NJK_STREAM_FACTORY as createNunjucksCompileStream,
} from './scripts/template.stream';

export {
  TPL_CFG_FILE_PATH,
} from './paths.const';

export {
  DocumentDescriptor,
  Path,
} from './interfaces';
