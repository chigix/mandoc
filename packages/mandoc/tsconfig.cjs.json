{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "sourceRoot": "../src",
    "module": "commonjs",
    "target": "es5",
    "outDir": "./dist/cjs",
  },
  "files": [
    "src/bin/mandoc.ts",
    "src/index.ts",
    "src/interfaces.ts",
    "src/paths.const.ts",
    "src/public_api.ts",
  ]
}