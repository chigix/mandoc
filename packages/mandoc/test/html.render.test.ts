import { fixture } from '@fixtures';
import * as fs from 'fs';
import { TPL_PKG_ALIASES } from 'mandoc/alias.const';
import { DocumentDescriptor } from 'mandoc/interfaces';
import { renderMarkdown } from 'mandoc/scripts/md2html.stream';
import { getConfig } from 'mandoc/scripts/template.config';
import * as path from 'path';
import * as through from 'through2';

function markdownStream(file: string) {
  return fs.createReadStream(file)
    .pipe(renderMarkdown({
      baseDir: path.dirname(file),
    }));
}

test('Basic RenderMarkdown', () => {
  markdownStream(fixture('ipsum.md'))
    .pipe(through.obj((report: DocumentDescriptor, enc, cb) => {
      expect(report.body).toMatchSnapshot();
      cb();
    }));
  markdownStream(fixture('keiyaku-test.md'))
    .pipe(through.obj((report: DocumentDescriptor, enc, cb) => {
      expect(report.body).toMatchSnapshot();
      cb();
    }));
});

test('Basic Template Rendering', done => {
  markdownStream(fixture('ipsum.md'))
    .pipe(
      getConfig(TPL_PKG_ALIASES.default, {
        paperWidth: 446,
        paperHeight: 446,
      }).renderers.njk.stream)
    .pipe(through((chunk: Buffer, enc, cb) => {
      expect(enc).toBe('buffer');
      expect(chunk.toString('utf8')).toMatchSnapshot();
      cb();
      done();
    }));
});
