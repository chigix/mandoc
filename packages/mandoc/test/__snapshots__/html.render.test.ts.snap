// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Template Rendering 1`] = `
"<h1>A First Level Header</h1>
<h2>A Second Level Header</h2>
<h3>A Third Level Header</h3>
<h4>A Fourth Level Header</h4>
<h5>A Fifth Level Header</h5>
<h6>A Sixth Level Header</h6>
<p>This is a paragraph. Cupcake ipsum dolor sit amet. Tiramisu liquorice gummies powder. Biscuit pastry pastry cake candy gummies carrot cake gingerbread. Gummi bears tart cupcake bonbon sugar plum biscuit. Icing marzipan tootsie roll. Pie gingerbread dragée wypas cookie. Faworki bonbon topping sweet soufflé jelly.</p>
<p>And another  paragraph. Jelly beans tootsie roll fruitcake toffee wypas chupa chups pudding. Soufflé muffin halvah powder sugar plum cake. Toffee cookie biscuit faworki pastry. Bonbon gummi bears topping caramels marshmallow. Chupa chups candy canes toffee. Powder powder candy canes pastry. Tootsie roll marshmallow powder.</p>
<p><em>These words are emphasized</em>
<em>Also these words are emphasized</em></p>
<p><strong>These words are strong emphasized</strong>
<strong>Also these words are strong emphasized</strong></p>
<blockquote>
<p>This is a very simple blockquote</p>
</blockquote>
<ul>
<li>This</li>
<li>is</li>
<li>an</li>
<li>unordered</li>
<li>list.</li>
</ul>
<p>And...</p>
<ol>
<li>This</li>
<li>is</li>
<li>an</li>
<li>ordered</li>
<li>list.</li>
</ol>
<p><a href=\\"http://example.com/\\">This is an example link</a>.
<a href=\\"http://example.com/\\" title=\\"¡Buenos días!\\">This is an example link with a title</a>.</p>
<p>This is a image
<img src=\\"http://placehold.it/350x150\\" alt=\\"Alt\\" title=\\"I'm an image\\"></p>
<p><code>this is a line of code</code></p>
<pre><code>And these are several lines of code. Pudding sesame snaps cupcake bonbon cupcake. Icing macaroon donut.

More code. Donut cheesecake chocolate cake. Tart pudding wafer dessert chocolate bar jelly-o apple pie dessert. Carrot cake apple pie pie liquorice oat cake lollipop pastry caramels. Jelly-o sesame snaps gummi bears.

And more. Sugar plum lemon drops chupa chups chocolate pastry. Faworki applicake applicake brownie topping liquorice chocolate liquorice icing. Cake pudding pudding cake candy sugar plum soufflé.\`
</code></pre>
<pre><code class=\\"language-js\\"><span class=\\"hljs-keyword\\">var</span> Remarkable = <span class=\\"hljs-built_in\\">require</span>(<span class=\\"hljs-string\\">'remarkable'</span>);
<span class=\\"hljs-keyword\\">var</span> hljs       = <span class=\\"hljs-built_in\\">require</span>(<span class=\\"hljs-string\\">'highlight.js'</span>) <span class=\\"hljs-comment\\">// https://highlightjs.org/</span>

<span class=\\"hljs-comment\\">// Actual default values</span>
<span class=\\"hljs-keyword\\">var</span> md = <span class=\\"hljs-keyword\\">new</span> Remarkable({
  <span class=\\"hljs-attr\\">highlight</span>: <span class=\\"hljs-function\\"><span class=\\"hljs-keyword\\">function</span> (<span class=\\"hljs-params\\">str, lang</span>) </span>{
    <span class=\\"hljs-keyword\\">if</span> (lang &amp;&amp; hljs.getLanguage(lang)) {
      <span class=\\"hljs-keyword\\">try</span> {
        <span class=\\"hljs-keyword\\">return</span> hljs.highlight(lang, str).value;
      } <span class=\\"hljs-keyword\\">catch</span> (err) {}
    }

    <span class=\\"hljs-keyword\\">try</span> {
      <span class=\\"hljs-keyword\\">return</span> hljs.highlightAuto(str).value;
    } <span class=\\"hljs-keyword\\">catch</span> (err) {}

    <span class=\\"hljs-keyword\\">return</span> <span class=\\"hljs-string\\">''</span>; <span class=\\"hljs-comment\\">// use external default escaping</span>
  }
});
</code></pre>
<p>And that's all ...
This is an <strong>extremely simple and basic</strong> guide. To understand a lot better this kind of dark magic called markdown, logically I recommend the official wiki of the creator. Go, run, jump to <a href=\\"http://daringfireball.net/projects/markdown/\\">the site of John Gruber</a>.</p>
"
`;

exports[`Basic Template Rendering 2`] = `
"<h1>建物賃貸借契約書</h1>
<p>賃借人（以下、「甲」という。）と賃貸人（以下、「乙」という。）は、乙はその所有する下記の建物（以下、「本件建物」という。）について、以下のとおり建物賃貸借契約を締結する。</p>
<ul>
<li>建物の所在: ◯◯県◯◯市◯◯</li>
</ul>
<h2>使用目的</h2>
<p>甲は本件物件の使用に際しては、居住にのみ使用し、その他の目的に使用してはならない。</p>
<h2>期間</h2>
<p>賃貸借期間は20XX年XX月XX日より20XX年XX月XX日までの満2ヵ年とする。但し、期間満了の6ヵ月前までに、本契約当事者のいずれかから書面による解約の通知がなされないときは、本契約は期間満了から更に1ヵ年自動的に延長されるものとし、以後も同様とする。</p>
<h2>賃料</h2>
<p>賃料は1ヶ月金XX万円とし、甲は、毎月末日までにその翌月分を乙の指定する銀行口座に振り込むものとする。但し、1ヵ月に満たない月の賃料は、日割計算とする。</p>
<h2>共益費</h2>
<p>甲は、共益費として賃借物件の使用上賃借人の負担となるべき諸経費を、賃料のほかに乙に支払わなければならない。</p>
<h2>敷金</h2>
<p>甲は、本契約締結時に、敷金として金XX万円を、本契約成立時に乙に預け入れる。ただし、敷金には利息は付さないものとし、本契約の解約による物件の明渡し時に甲に返還するものとする。</p>
<h2>賃借権の譲渡等の禁止</h2>
<p>甲は次の行為をしてはならない。</p>
<ol>
<li>賃借権の譲渡、本件建物の転貸</li>
<li>本件建物を第三者の使用に供すること</li>
</ol>
<h2>修繕費の負担</h2>
<p>乙は、本件建物が甲の使用目的に適するように維持保全に必要な修繕を行なう。</p>
<p>甲は、建具・造作、給排水施設、壁等、甲による日常の使用によって被る損耗について修理費用を負担する。</p>
<p>費用の負担につき疑義のあるときは、甲乙別途協議のうえ決定する。</p>
<hr>
<p>20XX年XX月XX日</p>
<h4>甲</h4>
<ul>
<li>住所:</li>
<li>氏名:</li>
</ul>
<h4>乙</h4>
<ul>
<li>住所:</li>
<li>氏名:</li>
</ul>
"
`;

exports[`Basic Template Rendering 3`] = `
"<!DOCTYPE html>
<html>
<head>
    <meta charset=\\"utf-8\\" />
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
    <title>Report</title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <style>
        @page {
            size: 446mm 446mm;
            margin: 2cm;
        }
        /*! normalize.css v8.0.1 | MIT License | github.com/necolas/normalize.css */

/* Document
   ========================================================================== */

/**
 * 1. Correct the line height in all browsers.
 * 2. Prevent adjustments of font size after orientation changes in iOS.
 */

html {
  line-height: 1.15; /* 1 */
  -webkit-text-size-adjust: 100%; /* 2 */
}

/* Sections
   ========================================================================== */

/**
 * Remove the margin in all browsers.
 */

body {
  margin: 0;
}

/**
 * Render the \`main\` element consistently in IE.
 */

main {
  display: block;
}

/**
 * Correct the font size and margin on \`h1\` elements within \`section\` and
 * \`article\` contexts in Chrome, Firefox, and Safari.
 */

h1 {
  font-size: 2em;
  margin: 0.67em 0;
}

/* Grouping content
   ========================================================================== */

/**
 * 1. Add the correct box sizing in Firefox.
 * 2. Show the overflow in Edge and IE.
 */

hr {
  box-sizing: content-box; /* 1 */
  height: 0; /* 1 */
  overflow: visible; /* 2 */
}

/**
 * 1. Correct the inheritance and scaling of font size in all browsers.
 * 2. Correct the odd \`em\` font sizing in all browsers.
 */

pre {
  font-family: monospace, monospace; /* 1 */
  font-size: 1em; /* 2 */
}

/* Text-level semantics
   ========================================================================== */

/**
 * Remove the gray background on active links in IE 10.
 */

a {
  background-color: transparent;
}

/**
 * 1. Remove the bottom border in Chrome 57-
 * 2. Add the correct text decoration in Chrome, Edge, IE, Opera, and Safari.
 */

abbr[title] {
  border-bottom: none; /* 1 */
  text-decoration: underline; /* 2 */
  text-decoration: underline dotted; /* 2 */
}

/**
 * Add the correct font weight in Chrome, Edge, and Safari.
 */

b,
strong {
  font-weight: bolder;
}

/**
 * 1. Correct the inheritance and scaling of font size in all browsers.
 * 2. Correct the odd \`em\` font sizing in all browsers.
 */

code,
kbd,
samp {
  font-family: monospace, monospace; /* 1 */
  font-size: 1em; /* 2 */
}

/**
 * Add the correct font size in all browsers.
 */

small {
  font-size: 80%;
}

/**
 * Prevent \`sub\` and \`sup\` elements from affecting the line height in
 * all browsers.
 */

sub,
sup {
  font-size: 75%;
  line-height: 0;
  position: relative;
  vertical-align: baseline;
}

sub {
  bottom: -0.25em;
}

sup {
  top: -0.5em;
}

/* Embedded content
   ========================================================================== */

/**
 * Remove the border on images inside links in IE 10.
 */

img {
  border-style: none;
}

/* Forms
   ========================================================================== */

/**
 * 1. Change the font styles in all browsers.
 * 2. Remove the margin in Firefox and Safari.
 */

button,
input,
optgroup,
select,
textarea {
  font-family: inherit; /* 1 */
  font-size: 100%; /* 1 */
  line-height: 1.15; /* 1 */
  margin: 0; /* 2 */
}

/**
 * Show the overflow in IE.
 * 1. Show the overflow in Edge.
 */

button,
input { /* 1 */
  overflow: visible;
}

/**
 * Remove the inheritance of text transform in Edge, Firefox, and IE.
 * 1. Remove the inheritance of text transform in Firefox.
 */

button,
select { /* 1 */
  text-transform: none;
}

/**
 * Correct the inability to style clickable types in iOS and Safari.
 */

button,
[type=\\"button\\"],
[type=\\"reset\\"],
[type=\\"submit\\"] {
  -webkit-appearance: button;
}

/**
 * Remove the inner border and padding in Firefox.
 */

button::-moz-focus-inner,
[type=\\"button\\"]::-moz-focus-inner,
[type=\\"reset\\"]::-moz-focus-inner,
[type=\\"submit\\"]::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

/**
 * Restore the focus styles unset by the previous rule.
 */

button:-moz-focusring,
[type=\\"button\\"]:-moz-focusring,
[type=\\"reset\\"]:-moz-focusring,
[type=\\"submit\\"]:-moz-focusring {
  outline: 1px dotted ButtonText;
}

/**
 * Correct the padding in Firefox.
 */

fieldset {
  padding: 0.35em 0.75em 0.625em;
}

/**
 * 1. Correct the text wrapping in Edge and IE.
 * 2. Correct the color inheritance from \`fieldset\` elements in IE.
 * 3. Remove the padding so developers are not caught out when they zero out
 *    \`fieldset\` elements in all browsers.
 */

legend {
  box-sizing: border-box; /* 1 */
  color: inherit; /* 2 */
  display: table; /* 1 */
  max-width: 100%; /* 1 */
  padding: 0; /* 3 */
  white-space: normal; /* 1 */
}

/**
 * Add the correct vertical alignment in Chrome, Firefox, and Opera.
 */

progress {
  vertical-align: baseline;
}

/**
 * Remove the default vertical scrollbar in IE 10+.
 */

textarea {
  overflow: auto;
}

/**
 * 1. Add the correct box sizing in IE 10.
 * 2. Remove the padding in IE 10.
 */

[type=\\"checkbox\\"],
[type=\\"radio\\"] {
  box-sizing: border-box; /* 1 */
  padding: 0; /* 2 */
}

/**
 * Correct the cursor style of increment and decrement buttons in Chrome.
 */

[type=\\"number\\"]::-webkit-inner-spin-button,
[type=\\"number\\"]::-webkit-outer-spin-button {
  height: auto;
}

/**
 * 1. Correct the odd appearance in Chrome and Safari.
 * 2. Correct the outline style in Safari.
 */

[type=\\"search\\"] {
  -webkit-appearance: textfield; /* 1 */
  outline-offset: -2px; /* 2 */
}

/**
 * Remove the inner padding in Chrome and Safari on macOS.
 */

[type=\\"search\\"]::-webkit-search-decoration {
  -webkit-appearance: none;
}

/**
 * 1. Correct the inability to style clickable types in iOS and Safari.
 * 2. Change font properties to \`inherit\` in Safari.
 */

::-webkit-file-upload-button {
  -webkit-appearance: button; /* 1 */
  font: inherit; /* 2 */
}

/* Interactive
   ========================================================================== */

/*
 * Add the correct display in Edge, IE 10+, and Firefox.
 */

details {
  display: block;
}

/*
 * Add the correct display in all browsers.
 */

summary {
  display: list-item;
}

/* Misc
   ========================================================================== */

/**
 * Add the correct display in IE 10+.
 */

template {
  display: none;
}

/**
 * Add the correct display in IE 10.
 */

[hidden] {
  display: none;
}

        body .count-from-h1 {
  counter-reset: h1counter;
}
body .count-from-h1 h1 {
  counter-reset: h2counter;
}
body .count-from-h1 h1:before {
  counter-increment: h1counter;
  content: counter(h1counter)'.\\\\0000a0\\\\0000a0';
}
body .count-from-h1 h1.no-count:before {
  content: none;
  counter-increment: none;
}
body .count-from-h1 h2 {
  counter-reset: h3counter;
}
body .count-from-h1 h2:before {
  counter-increment: h2counter;
  content: counter(h1counter) '.' counter(h2counter)'.\\\\0000a0\\\\0000a0';
}
body .count-from-h1 h2.no-count:before {
  content: none;
  counter-increment: none;
}
body .count-from-h1 h3 {
  counter-reset: h4counter;
}
body .count-from-h1 h3:before {
  counter-increment: h3counter;
  content: counter(h1counter) '.' counter(h2counter) '.' counter(h3counter)'.\\\\0000a0\\\\0000a0';
}
body .count-from-h1 h3.no-count:before {
  content: none;
  counter-increment: none;
}
body .count-from-h1 h4 {
  counter-reset: h5counter;
}
body .count-from-h1 h4:before {
  counter-increment: h4counter;
  content: counter(h1counter) '.' counter(h2counter) '.' counter(h3counter) '.' counter(h4counter)'.\\\\0000a0\\\\0000a0';
}
body .count-from-h1 h4.no-count:before {
  content: none;
  counter-increment: none;
}
body .count-from-h1 h5 {
  counter-reset: h6counter;
}
body .count-from-h1 h5:before {
  counter-increment: h5counter;
  content: counter(h1counter) '.' counter(h2counter) '.' counter(h3counter) '.' counter(h4counter) '.' counter(h5counter)'.\\\\0000a0\\\\0000a0';
}
body .count-from-h1 h5.no-count:before {
  content: none;
  counter-increment: none;
}
body .count-from-h1 h6 {
  counter-reset: h7counter;
}
body .count-from-h1 h6:before {
  counter-increment: h6counter;
  content: counter(h1counter) '.' counter(h2counter) '.' counter(h3counter) '.' counter(h4counter) '.' counter(h5counter) '.' counter(h6counter)'.\\\\0000a0\\\\0000a0';
}
body .count-from-h1 h6.no-count:before {
  content: none;
  counter-increment: none;
}
body .count-from-h2 {
  counter-reset: h2counter;
}
body .count-from-h2 h2 {
  counter-reset: h3counter;
}
body .count-from-h2 h2:before {
  counter-increment: h2counter;
  content: counter(h2counter)'.\\\\0000a0\\\\0000a0';
}
body .count-from-h2 h2.no-count:before {
  content: none;
  counter-increment: none;
}
body .count-from-h2 h3 {
  counter-reset: h4counter;
}
body .count-from-h2 h3:before {
  counter-increment: h3counter;
  content: counter(h2counter) '.' counter(h3counter)'.\\\\0000a0\\\\0000a0';
}
body .count-from-h2 h3.no-count:before {
  content: none;
  counter-increment: none;
}
body .count-from-h2 h4 {
  counter-reset: h5counter;
}
body .count-from-h2 h4:before {
  counter-increment: h4counter;
  content: counter(h2counter) '.' counter(h3counter) '.' counter(h4counter)'.\\\\0000a0\\\\0000a0';
}
body .count-from-h2 h4.no-count:before {
  content: none;
  counter-increment: none;
}
body .count-from-h2 h5 {
  counter-reset: h6counter;
}
body .count-from-h2 h5:before {
  counter-increment: h5counter;
  content: counter(h2counter) '.' counter(h3counter) '.' counter(h4counter) '.' counter(h5counter)'.\\\\0000a0\\\\0000a0';
}
body .count-from-h2 h5.no-count:before {
  content: none;
  counter-increment: none;
}
body .count-from-h2 h6 {
  counter-reset: h7counter;
}
body .count-from-h2 h6:before {
  counter-increment: h6counter;
  content: counter(h2counter) '.' counter(h3counter) '.' counter(h4counter) '.' counter(h5counter) '.' counter(h6counter)'.\\\\0000a0\\\\0000a0';
}
body .count-from-h2 h6.no-count:before {
  content: none;
  counter-increment: none;
}
/** For screen preview **/
@media screen {
  article {
    width: 100%;
    max-width: 960px;
    margin: 5px auto;
  }
}
/** Fix for Chrome issue #273306 **/
@media print {
  body.A3.landscape {
    width: 420mm;
  }
  body.A3,
  body.A4.landscape {
    width: 297mm;
  }
  body.A4,
  body.A5.landscape {
    width: 210mm;
  }
  body.A5 {
    width: 148mm;
  }
  body.letter,
  body.legal {
    width: 216mm;
  }
  body.letter.landscape {
    width: 280mm;
  }
  body.legal.landscape {
    width: 357mm;
  }
}
article {
  font-size: 12pt;
  font-family: -apple-system, \\"BlinkMacSystemFont\\", \\"Helvetica Neue\\", Helvetica, \\"Arial\\", \\"ヒラギノ角ゴ ProN W3\\", \\"Hiragino Kaku Gothic ProN\\", \\"メイリオ\\", Meiryo, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\", \\"Noto Color Emoji\\";
  hyphens: auto;
  /* turn on hyphenation */
  text-align: justify;
  /* and justification */
}
article.print-two-column {
  columns: 2;
  column-gap: 0.25in;
}
article section.lead {
  column-span: all;
  text-align: center;
  margin: 2em 0;
  font-style: italic;
}
article section.lead h1 {
  font-style: normal;
}

    </style>
    <link rel=\\"stylesheet\\" href=\\"file:///C:\\\\Users\\\\chigi\\\\Projects\\\\mandoc\\\\packages\\\\tpl-paper\\\\node_modules\\\\highlight.js\\\\styles\\\\atelier-lakeside-light.css\\">
    <script type=\\"text/x-mathjax-config\\">
    MathJax.Hub.Config({
        tex2jax: {inlineMath: [['$','$'], ['\\\\\\\\(','\\\\\\\\)']]}
    });
    </script>
</head>
<body>
    <article class=\\" count-from-h2\\">
        <section id=\\"jpikr1tr\\" class=\\"lead\\"><h1>A First Level Header</h1></section>
<h2>A Second Level Header</h2>
<h3>A Third Level Header</h3>
<h4>A Fourth Level Header</h4>
<h5>A Fifth Level Header</h5>
<h6>A Sixth Level Header</h6>
<p>This is a paragraph. Cupcake ipsum dolor sit amet. Tiramisu liquorice gummies powder. Biscuit pastry pastry cake candy gummies carrot cake gingerbread. Gummi bears tart cupcake bonbon sugar plum biscuit. Icing marzipan tootsie roll. Pie gingerbread drag&#xE9;e wypas cookie. Faworki bonbon topping sweet souffl&#xE9; jelly.</p>
<p>And another  paragraph. Jelly beans tootsie roll fruitcake toffee wypas chupa chups pudding. Souffl&#xE9; muffin halvah powder sugar plum cake. Toffee cookie biscuit faworki pastry. Bonbon gummi bears topping caramels marshmallow. Chupa chups candy canes toffee. Powder powder candy canes pastry. Tootsie roll marshmallow powder.</p>
<p><em>These words are emphasized</em>
<em>Also these words are emphasized</em></p>
<p><strong>These words are strong emphasized</strong>
<strong>Also these words are strong emphasized</strong></p>
<blockquote>
<p>This is a very simple blockquote</p>
</blockquote>
<ul>
<li>This</li>
<li>is</li>
<li>an</li>
<li>unordered</li>
<li>list.</li>
</ul>
<p>And...</p>
<ol>
<li>This</li>
<li>is</li>
<li>an</li>
<li>ordered</li>
<li>list.</li>
</ol>
<p><a href=\\"http://example.com/\\">This is an example link</a>.
<a href=\\"http://example.com/\\" title=\\"&#xA1;Buenos d&#xED;as!\\">This is an example link with a title</a>.</p>
<p>This is a image
<img src=\\"http://placehold.it/350x150\\" alt=\\"Alt\\" title=\\"I&apos;m an image\\"></p>
<p><code>this is a line of code</code></p>
<pre><code>And these are several lines of code. Pudding sesame snaps cupcake bonbon cupcake. Icing macaroon donut.

More code. Donut cheesecake chocolate cake. Tart pudding wafer dessert chocolate bar jelly-o apple pie dessert. Carrot cake apple pie pie liquorice oat cake lollipop pastry caramels. Jelly-o sesame snaps gummi bears.

And more. Sugar plum lemon drops chupa chups chocolate pastry. Faworki applicake applicake brownie topping liquorice chocolate liquorice icing. Cake pudding pudding cake candy sugar plum souffl&#xE9;.\`
</code></pre>
<pre><code class=\\"language-js\\"><span class=\\"hljs-keyword\\">var</span> Remarkable = <span class=\\"hljs-built_in\\">require</span>(<span class=\\"hljs-string\\">&apos;remarkable&apos;</span>);
<span class=\\"hljs-keyword\\">var</span> hljs       = <span class=\\"hljs-built_in\\">require</span>(<span class=\\"hljs-string\\">&apos;highlight.js&apos;</span>) <span class=\\"hljs-comment\\">// https://highlightjs.org/</span>

<span class=\\"hljs-comment\\">// Actual default values</span>
<span class=\\"hljs-keyword\\">var</span> md = <span class=\\"hljs-keyword\\">new</span> Remarkable({
  <span class=\\"hljs-attr\\">highlight</span>: <span class=\\"hljs-function\\"><span class=\\"hljs-keyword\\">function</span> (<span class=\\"hljs-params\\">str, lang</span>) </span>{
    <span class=\\"hljs-keyword\\">if</span> (lang &amp;&amp; hljs.getLanguage(lang)) {
      <span class=\\"hljs-keyword\\">try</span> {
        <span class=\\"hljs-keyword\\">return</span> hljs.highlight(lang, str).value;
      } <span class=\\"hljs-keyword\\">catch</span> (err) {}
    }

    <span class=\\"hljs-keyword\\">try</span> {
      <span class=\\"hljs-keyword\\">return</span> hljs.highlightAuto(str).value;
    } <span class=\\"hljs-keyword\\">catch</span> (err) {}

    <span class=\\"hljs-keyword\\">return</span> <span class=\\"hljs-string\\">&apos;&apos;</span>; <span class=\\"hljs-comment\\">// use external default escaping</span>
  }
});
</code></pre>
<p>And that&apos;s all ...
This is an <strong>extremely simple and basic</strong> guide. To understand a lot better this kind of dark magic called markdown, logically I recommend the official wiki of the creator. Go, run, jump to <a href=\\"http://daringfireball.net/projects/markdown/\\">the site of John Gruber</a>.</p>

        
    </article>
    <script type=\\"text/javascript\\" src=\\"file:///C:\\\\Users\\\\chigi\\\\Projects\\\\mandoc\\\\packages\\\\tpl-paper\\\\node_modules\\\\mathjax\\\\MathJax.js?config=TeX-AMS_HTML-full\\"></script>
</body>
</html>"
`;
